text/0000-oracles.md

```diff
@@ -0,0 +1,136 @@
+| Status  | Stakeholders | Outcome | Due date |   Owner   |
+| :-----: | :----------: | :-----: | :------: | :-------: |
+| Started |              |         |          | @arjentix |
+
+
+
+## Background
+
+"Oracle" is a common name for different technologies which basically do one thing – bring off-chain data on chain. The possibilities are almost limitless:
+
+1. Real wold currencies
+2. Random numbers generation
+3. Timestamp retrieving
+4. Events in other blockchains
+5. Election results
+6. Geo-data i.e. to track delivery chains
+7. Complex calculations
+8. Weather data
+9. Statistical data
+10. The central bank's key rate
+
+Obviously integrating Oracles into Iroha 2 based blockchains will expand the horizon on possibilities for our users. By providing a clear way to integrate your WASM Trigger with Oracles we encourage people to use Iroha 2 in really complex systems and bring their ideas to life.
+
+## Problem
+
+We need to add Oracles support into Iroha 2. Oracles should be a flexible part of a blockchain and should not be hardcoded anywhere on our side. There should be a clear mechanism to interact with them and respective documentation.
+
+### Example scenarios
+
+ Scenario 1:
+
+1. End-user interacts with a trigger to buy some asset which is tied to a real world, i.e. gold, for some other asset A;
+2. Trigger calls Oracle 1 to get gold price in dollars, and then calls Oracle 2 to get asset A price in dollars;
+3. Trigger computes the result and transfers money to the buyer.
+
+Scenario 2:
+
+1. Weather sensors have registered a hurricane
+2. Trigger has been activated to transfer insurance benefits to the clients
+
+ Scenario 3:
+
+1. End-user interacts with a trigger, requesting some permission token to be given to their account. This permission can be given only to account which corresponds to University "X" students.
+2. Trigger calls Oracle of University "X" to ensure that account corresponds to a student
+3. On success trigger sends permission token to the account
+
+## Solution
+
+Proposed solution is based on already built parts of Iroha and does not require much efforts to implement.
+It consists of the two parts:
+
+1. `EmitOracleRequest` Iroha Special Instruction;
+2. `OracleRequest` event type.
```

Comment:
I think it could be further generalised with probably higher expressiveness:

- Instruction to emit an event with some payload: `EmitCustomEvent(Payload)`
- An event type which is emitted in case of instruction execution: `CustomEvent(Payload)`

This removes the semantic limitation of "oracles", but is totally the same. Also, it clearly transmits that Iroha might be used as a pub-sub event bus.

Reply:
I don't think that functionally it remains the same. The problem in your approach is filtering. Oracles will probably be a hot path, so a proper event filter implemented on our side might help a lot. `OracleRequest` will probably contain something like `query_id: String` to easier filtering

Reply:
From your words, it will be a *tagged* pub-sub event bus, with `EmitOracleRequest(Id, Payload)` (with `Id` being a "query" id, or let's just call it a tag of the event) and `OracleRequest` event filters by the `Id` of the emitted event. I agree that filtering is important here, however, I wonder if a string identifier will suffice.

Still, this is a pub-sub event bus. Therefore, its functionality might go out of scope of Oracles use case, especially if Iroha itself might be triggered by this kind of events. Why should its naming be restricted specifically to the Oracles use case?

Reply:
Hm, I probably meant `oracle_id` instead of `query_id`, sorry.

Reply:
I still don't quite understand your idea. We already have pub-sub event bus. I suggest to just add another event type for it.

To me it doesn't look like enabled any custom event type. This seems like another task which needs clarification. For example, I don't see any use-cases for that.

However that might be useful l when we'll have RISC architecture. So that `Executor` should be able to emit any event defined inside it. Before implementing RISC architecture I don't see how to use `CustomEvent`

Reply:
So, you are suggesting to add another event type specifically for Oracles. And it doesn't look for you like it enables support for any custom event types. The event type has a string id, and do not have any restrictions on payload. As for custom events, you don't see a use case for it, and you see Oracles events as something different from it. Do I understand correctly so far?

If so, I don't see a difference between new Oracles events and just custom events with id and payload. Thus, I don't see why it makes sense to use "Oracles" in its naming. Why?

Reply:
> well, partially I see why it makes sense: according to the RFC, any off-chain actor that uses custom events is an oracle by definition (AFAIU). Therefore, custom events are oracle events by definition as well.

Yes, I think that's it. Don't you agree?

Reply:
Yeah, agreed.